Tasks/SonarQubePostTest/PRCA/PostComments-Module.psm1

```diff
@@ -181,39 +299,190 @@ function GetPullRequestId
+}
+
+function FilterCommentsByPath
+{
+    param ([Array]$messages)
+    
+    $modifiedFilesInPr = GetModifiedFilesInPR
+    Write-Verbose "Files changed in this PR: $modifiedFilesInPr"
+    
+    $countBefore = $messages.Count
+    $messages = $messages | Where-Object {$modifiedFilesInPr.Contains($_.RelativePath)}
+    $commentsFiltered = $countBefore - $messages.Count 
+    
+    Write-Host "$commentsFiltered message(s) were filtered because they do not belong to files that were changed in this PR"
+    
+    return $messages
+}
+
+#
+# A message is said to be pre-existing if:
+# 1. a comment with the same content exists 
+# 2. and it belongs to the same file 
+# 3. and the thread was created by the same logic, i.e. the same commentSource 
+#
+# Remark: comments move arround so Line cannot be used. 
+#
+function FilterPreExistingComments
+{
+     param ([Array]$messages, 
+     [ValidateNotNullOrEmpty()][string]$messageSource, 
+     [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionThread]]$existingThreads,
+     [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionComment]]$existingComments)
+     
+     $sw = new-object "Diagnostics.Stopwatch"
+     $sw.Start();
+     
+     $countBefore = $messages.Count
+     $messages = $messages | Where-Object { !(MessageHasMatchingComments $_)}
+     $commentsFiltered = $countBefore - $messages.Count 
+     
+     Write-Host "$commentsFiltered message(s) were filtered because they were already present"
+     Write-Verbose "Filtering out $($existingComments.Count) existing comments took $($sw.ElapsedMilliseconds) ms"
+     
+     return $messages
+}
+
+function MessageHasMatchingComments
+{
+    param ([ValidateNotNull()][PSObject]$message)
+    
+    $matchingComments = GetMatchingComments $_ $messageSource $existingThreads $existingComments
+    
+    return ($matchingComments.Count -gt 0)
+}
+
+#TODO: can be optimized by using a map of <Thread,List<Comments>> instead of 2 flat lists
+function GetMatchingComments
+{
+     param ([ValidateNotNull()][PSObject]$message, 
+     [ValidateNotNullOrEmpty()][string]$messageSource, 
+     [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionThread]]$existingThreads,
+     [System.Collections.Generic.List[Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionComment]]$existingComments)
+     
+     $resultList = @()
+     
+     # select threads that are not "fixed", that point to the same file and have been marked with the given comment source
+     $matchingThreads = $existingThreads | Where-Object {
+            ($_ -ne $null) -and
+            (ThreadMatchesCommentSource $_ $messageSource) -and
+            (ThreadMatchesItemPath $_ $message.RelativePath) -and 
+            ($_.Status -ne [Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionStatus]::Fixed)}
```

Comment:
Minor optimisation: assuming PowerShell short-circuits logical operations, you could moved the status check to after the null check - avoids calling the two string-comparison methods